operator/crd.go

```diff
@@ -0,0 +1,53 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"time"
+
+	"github.com/cilium/cilium/pkg/controller"
+
+	"github.com/pkg/errors"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func getCRD(client clientset.Interface, name string) error {
+	log.Debugf("Getting CRD %s from Kubernetes apiserver", name)
+	if _, err := client.ApiextensionsV1beta1().CustomResourceDefinitions().Get(context.TODO(), name, metav1.GetOptions{}); err != nil {
+		log.WithError(err).Debugf("Could not get CRD %s from Kubernetes apiserver, retrying", name)
+		return err
+	}
+	log.Debugf("Check of CRD %s successful", name)
+	return nil
+}
+
+// waitForCRD waits for the given CRD to be available until the timeout defined
+// by CRDRetry. Returns an error when timeout exceeded.
+func waitForCRD(client clientset.Interface, name string) error {
+	ctrlMgr := controller.NewManager()
+	ctrlName := "wait-for-crd-" + name
+	ctrlMgr.UpdateController(ctrlName, controller.ControllerParams{
+		RunInterval: 15 * time.Second,
+		DoFunc: func(ctx context.Context) error {
+			return getCRD(client, name)
+		},
+	})
+	if err := ctrlMgr.RemoveControllerAndWait(ctrlName); err != nil {
+		return errors.Wrapf(err, "Could not wait for CRD %s status", name)
+	}
+	return nil
+}
```

Comment:
I'm a little lost by this function. It seems it will create a controller and then will remove it. Is this intended?

Reply:
Yes - my intention is to initially wait until we can get the CRD and then remove the contoller once we get it successfully. I see no need to have that controller running afterwards.

Reply:
Right but line 43 is non blocking, i.e. the controller will be updated and removed immediately.

Reply:
`RemoveControllerAndWait` will wait until the controller has completed and as far as I understand, it's blocking. That's my log from cilium-operator:

```
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="Getting CRD ciliumnetworkpolicies.cilium.io from Kubernetes apiserver" subsys=cilium-operator
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="Controller func execution time: 1.336µs" name=kvstore-sync-store-cilium/state/cnpstatuses/v2 subsys=controller uuid=5a5870de-991e-11ea-820a-08002737a2e9
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="Controller run succeeded; waiting for next controller update or stop" name=kvstore-sync-store-cilium/state/cnpstatuses/v2 subsys=controller uuid=5a5870de-991e-11ea-820a-08002737a2e9
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="Starting watcher..." config=/var/lib/cilium/etcd-config.yml endpoints="[https://192.168.33.11:2379]" subsys=kvstore watcher=cnpStatusWatcher
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="Distributed lock successful, etcd has quorum" config=/var/lib/cilium/etcd-config.yml endpoints="[https://192.168.33.11:2379]" subsys=kvstore
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="List response from etcd len=0: &{Header:cluster_id:6438720328480595214 member_id:1326427008500738502 revision:408 raft_term:2  Kvs:[] More:false Count:0}" config=/var/lib/cilium/etcd-config.yml endpoints="[https://192.168.33.11:2379]" prefix=cilium/state/cnpstatuses/v2 subsys=kvstore watcher=cnpStatusWatcher
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="Starting to watch a prefix" config=/var/lib/cilium/etcd-config.yml endpoints="[https://192.168.33.11:2379]" prefix=cilium/state/cnpstatuses/v2 revision=409 subsys=kvstore watcher=cnpStatusWatcher
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="Check of CRD ciliumnetworkpolicies.cilium.io successful" subsys=cilium-operator
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="Controller func execution time: 55.037927ms" name=wait-for-crd-ciliumnetworkpolicies.cilium.io subsys=controller uuid=5a58728f-991e-11ea-820a-08002737a2e9
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="Shutting down controller" name=wait-for-crd-ciliumnetworkpolicies.cilium.io subsys=controller uuid=5a58728f-991e-11ea-820a-08002737a2e9
```

Which seems to prove that the controller was blocked until it finished its successful iteration.

But on the other hand, it's the first time when I ever use the controller, so I might be wrong.

Reply:
Anyway, if I'm wrong, I'm open for suggestions how should I use the controller to wait until it's first successful run.

Reply:
Or no, sorry, I'm wrong. There is no guarantee that there was at least one successful iteration of controller when using `RemoveControllerAndWait`. I need to come up with some solution which will ensure at least one success.

Reply:
Should be fixed now by 62ce6814d6429b79d2ce0a351a123f28bac68124